File: python/src/contriwork_notifications/adapters/pushover.py
"""Pushover adapter — HTTPS POST to api.pushover.net."""

from __future__ import annotations

import httpx

from ..adapter import AdapterDeliverResult, AdapterStatus
from ..errors import ErrorCode
from ..payload import Payload
from ..severity import Severity

_DEFAULT_API_URL = "https://api.pushover.net/1/messages.json"
_DEFAULT_TIMEOUT_S = 10.0

# CONTRACT.md §Security invariants: detail strings never echo tokens or
# response bodies. Generic strings only.
_GENERIC_FAILURE = "Notification delivery failed"
_GENERIC_AUTH = "Authentication rejected"


class PushoverAdapter:
    """Pushover delivery via HTTPS POST.

    Severity → Pushover priority mapping:
      * DEBUG    → -1 (silent / badge-only)
      * INFO     →  0
      * WARN     →  0
      * ERROR    →  ``error_priority`` (default 0; pass 1 to escalate)
      * CRITICAL →  1 (high priority; bypasses Pushover's own quiet hours)

    Silent mode (active quiet hours per the client config): priority is
    forced to -1. Severities listed in the client's ``bypass_severities``
    never reach silent mode in the first place.
    """

    def __init__(
        self,
        user_key: str,
        app_token: str,
        *,
        name: str = "pushover",
        error_priority: int = 0,
        api_url: str = _DEFAULT_API_URL,
        timeout_seconds: float = _DEFAULT_TIMEOUT_S,
    ) -> None:
        if error_priority not in (0, 1):
            raise ValueError("error_priority must be 0 or 1")
        self._user_key = user_key
        self._app_token = app_token
        self.name = name
        self._error_priority = error_priority
        self._api_url = api_url
        self._timeout = timeout_seconds

    async def is_available(self) -> bool:
        return bool(self._user_key and self._app_token)

    async def deliver(
        self,
        severity: Severity,
        payload: Payload,
        silent: bool,
    ) -> AdapterDeliverResult:
        priority = -1 if silent else self._priority_for(severity)
        data: dict[str, str | int] = {
            "token": self._app_token,
            "user": self._user_key,
            "title": payload.title,
            "message": payload.body,
            "priority": priority,
        }
        if payload.url is not None:
            data["url"] = payload.url
        if payload.url_title is not None:
            data["url_title"] = payload.url_title

        try:
            async with httpx.AsyncClient(timeout=self._timeout) as client:
                response = await client.post(self._api_url, data=data)
        except httpx.TimeoutException:
            return AdapterDeliverResult(
                status=AdapterStatus.RETRIABLE_FAILURE,
                error_code=ErrorCode.TIMEOUT,
                detail=_GENERIC_FAILURE,
            )
        except httpx.HTTPError:
            return AdapterDeliverResult(
                status=AdapterStatus.RETRIABLE_FAILURE,
                error_code=ErrorCode.UPSTREAM_ERROR,
                detail=_GENERIC_FAILURE,
            )

        return _map_response(response.status_code)

    def _priority_for(self, severity: Severity) -> int:
        if severity == Severity.DEBUG:
            return -1
        if severity == Severity.CRITICAL:
            return 1
        if severity == Severity.ERROR:
            return self._error_priority
        return 0  # INFO, WARN


def _map_response(status_code: int) -> AdapterDeliverResult:
    if 200 <= status_code < 300:
        return AdapterDeliverResult(status=AdapterStatus.DELIVERED)
    if status_code == 429:
        return AdapterDeliverResult(
            status=AdapterStatus.RETRIABLE_FAILURE,
            error_code=ErrorCode.RATE_LIMITED,
            detail=_GENERIC_FAILURE,
        )
    if status_code in (401, 403):
        return AdapterDeliverResult(
            status=AdapterStatus.PERMANENT_FAILURE,
            error_code=ErrorCode.AUTH_FAILED,
            detail=_GENERIC_AUTH,
        )
    if 400 <= status_code < 500:
        return AdapterDeliverResult(
            status=AdapterStatus.PERMANENT_FAILURE,
            error_code=ErrorCode.ADAPTER_INVALID_PAYLOAD,
            detail=_GENERIC_FAILURE,
        )
    return AdapterDeliverResult(
        status=AdapterStatus.RETRIABLE_FAILURE,
        error_code=ErrorCode.UPSTREAM_ERROR,
        detail=_GENERIC_FAILURE,
    )

File: typescript/src/adapters/pushover.ts
import type { Adapter, AdapterDeliverResult } from "../adapter";
import { AdapterStatus } from "../adapter";
import { ErrorCode } from "../errors";
import type { Payload } from "../payload";
import { Severity } from "../severity";

const DEFAULT_API_URL = "https://api.pushover.net/1/messages.json";
const DEFAULT_TIMEOUT_MS = 10_000;
const GENERIC_FAILURE = "Notification delivery failed";
const GENERIC_AUTH = "Authentication rejected";

export interface PushoverOptions {
  readonly userKey: string;
  readonly appToken: string;
  readonly name?: string;
  readonly errorPriority?: 0 | 1;
  readonly apiUrl?: string;
  readonly timeoutMs?: number;
  readonly fetchImpl?: typeof fetch;
}

/**
 * Pushover delivery via HTTPS POST to api.pushover.net/1/messages.json.
 *
 * Severity → Pushover priority mapping:
 *   - Debug    → -1 (silent / badge-only)
 *   - Info     →  0
 *   - Warn     →  0
 *   - Error    →  errorPriority (default 0; pass 1 to escalate)
 *   - Critical →  1 (high priority)
 *
 * Silent mode: priority forced to -1.
 */
export class PushoverAdapter implements Adapter {
  readonly name: string;
  private readonly userKey: string;
  private readonly appToken: string;
  private readonly errorPriority: 0 | 1;
  private readonly apiUrl: string;
  private readonly timeoutMs: number;
  private readonly fetchImpl: typeof fetch;

  constructor(options: PushoverOptions) {
    this.userKey = options.userKey;
    this.appToken = options.appToken;
    this.name = options.name ?? "pushover";
    this.errorPriority = options.errorPriority ?? 0;
    this.apiUrl = options.apiUrl ?? DEFAULT_API_URL;
    this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.fetchImpl = options.fetchImpl ?? globalThis.fetch.bind(globalThis);
  }

  async isAvailable(): Promise<boolean> {
    return Boolean(this.userKey) && Boolean(this.appToken);
  }

  async deliver(
    severity: Severity,
    payload: Payload,
    silent: boolean,
  ): Promise<AdapterDeliverResult> {
    const priority = silent ? -1 : this.priorityFor(severity);
    const body = new URLSearchParams();
    body.set("token", this.appToken);
    body.set("user", this.userKey);
    body.set("title", payload.title);
    body.set("message", payload.body);
    body.set("priority", String(priority));
    if (payload.url !== undefined) body.set("url", payload.url);
    if (payload.urlTitle !== undefined) body.set("url_title", payload.urlTitle);

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    let response: Response;
    try {
      response = await this.fetchImpl(this.apiUrl, {
        method: "POST",
        body,
        signal: controller.signal,
      });
    } catch {
      if (controller.signal.aborted) {
        return {
          status: AdapterStatus.RetriableFailure,
          errorCode: ErrorCode.Timeout,
          detail: GENERIC_FAILURE,
        };
      }
      return {
        status: AdapterStatus.RetriableFailure,
        errorCode: ErrorCode.UpstreamError,
        detail: GENERIC_FAILURE,
      };
    } finally {
      clearTimeout(timer);
    }

    return mapResponse(response.status);
  }

  private priorityFor(severity: Severity): number {
    if (severity === Severity.Debug) return -1;
    if (severity === Severity.Critical) return 1;
    if (severity === Severity.Error) return this.errorPriority;
    return 0;
  }
}

function mapResponse(statusCode: number): AdapterDeliverResult {
  if (statusCode >= 200 && statusCode < 300) {
    return { status: AdapterStatus.Delivered };
  }
  if (statusCode === 429) {
    return {
      status: AdapterStatus.RetriableFailure,
      errorCode: ErrorCode.RateLimited,
      detail: GENERIC_FAILURE,
    };
  }
  if (statusCode === 401 || statusCode === 403) {
    return {
      status: AdapterStatus.PermanentFailure,
      errorCode: ErrorCode.AuthFailed,
      detail: GENERIC_AUTH,
    };
  }
  if (statusCode >= 400 && statusCode < 500) {
    return {
      status: AdapterStatus.PermanentFailure,
      errorCode: ErrorCode.AdapterInvalidPayload,
      detail: GENERIC_FAILURE,
    };
  }
  return {
    status: AdapterStatus.RetriableFailure,
    errorCode: ErrorCode.UpstreamError,
    detail: GENERIC_FAILURE,
  };
}
